### PR TITLE
Issue #24717 Fixes Document number sequence with multiple docs open

### PR DESCRIPTION
**Pull request by andersdd**
Fixes Document number sequence with multiple docs open by different people at the same time and subsequent cancelling and re-opening causing document number sequence issues.

Will possibly cause document number gaps if gapless is not enabled.

Should also resolve for all document types.

### CODE REVIEW

**Comment by gilmoskowitz** [comment]
+1 but it would be very helpful to add a test for the `fetchnextnumber` and `releasenumber` functions. These functions have given us a lot of grief over the years and it would be nice to have unit tests to check for regressions.

**Comment by andersdd** [comment]
I merged the latest 4.9 code and the errors look to be related to tests outside of the changes I have made.

**Comment by discountscott** [comment]
Even with these changes, we still show the problem during testing. If Sales Rep 1 opens a Quote but doesn't enter any line items (just keeps it open for a bit). Sales rep 2 opens a new quote and then closes it. Sales rep 2 then opens a new quote. They both now have the same number. Whoever goes to save last or tries to add a line item will get the duplicate # error message.

**Comment by ssshyam** [comment]
Performed UAT: Observed that issue is not fixed for Quotes and the following db log error is generated on selecting to create second quote with same order number.

DB log Error:
Fri Jan 30 18:32:59 2015 ERROR:  duplicate key value violates unique constraint "quhead_quhead_number_key"
DETAIL:  Key (quhead_number)=(40017) already exists.
QPSQL: Unable to create query
INSERT INTO quhead (    quhead_id, quhead_number, quhead_cust_id,    quhead_custponumber, quhead_shipto_id,    quhead_billtoname, quhead_billtoaddress1,    quhead_billtoaddress2, quhead_billtoaddress3,    quhead_billtocity, quhead_billtostate, quhead_billtozip,    quhead_billtocountry,    quhead_shiptoname, quhead_shiptoaddress1,    quhead_shiptoaddress2, quhead_shiptoaddress3,    quhead_shiptocity, quhead_shiptostate, quhead_shiptozipcode,    quhead_shiptocountry,    quhead_quotedate, quhead_packdate,    quhead_salesrep_id, quhead_commission,    quhead_taxzone_id, quhead_terms_id,    quhead_shipvia, quhead_fob,    quhead_freight, quhead_calcfreight,    quhead_misc, quhead_misc_accnt_id, quhead_misc_descrip,    quhead_ordercomments, quhead_shipcomments,    quhead_prj_id, quhead_ophead_id, quhead_warehous_id,    quhead_curr_id, quhead_expire,    quhead_shipto_cntct_id,    quhead_shipto_cntct_honorific,    quhead_shipto_cntct_first_name,    quhead_shipto_cntct_middle,    quhead_shipto_cntct_last_name,    quhead_shipto_cntct_suffix,    quhead_shipto_cntct_phone,    quhead_shipto_cntct_title,    quhead_shipto_cntct_fax,    quhead_shipto_cntct_email,    quhead_billto_cntct_id,    quhead_billto_cntct_honorific,    quhead_billto_cntct_first_name,    quhead_billto_cntct_middle,    quhead_billto_cntct_last_name,    quhead_billto_cntct_suffix,    quhead_billto_cntct_phone,    quhead_billto_cntct_title,    quhead_billto_cntct_fax,    quhead_billto_cntct_email,    quhead_status,    quhead_shipzone_id, quhead_saletype_id)    VALUES (    139, '40017', 95,    '', 31,    'Tremendous Toys Incorporated', 'Tremendous Toys Inc.',    '101 Toys Place', '',    'Walnut Hills', 'VA', '22209',    'United States',    'Olde Towne Store 1', 'Olde Towne Toys Store 1',    '1 Duke Street', '',    'Alexandria', 'VA', '22201',    'United States',    '2015-01-30', '2015-01-30',    32, 0,    2, 42,    'UPS-GROUND-UPS Ground', '',    0, TRUE,    0, NULL, '',    '', 'Packing List Note:  No special handling or delivery instructions for 
deliveries to STORE1.
',    NULL, NULL, 35,    1, NULL,    1,    'Mr.',    'Mike',    '',    'Farley',    '',    '703-931-4269',    NULL,    '703-931-2212',    'mikef@ttoys.com',    1,    'Mr.',    'Mike',    '',    'Farley',    '',    '703-931-4269',    NULL,    '703-931-2212',    'mikef@ttoys.com',    'O',    35, 1)

**Comment by gpazo** [comment]
@andersdd can you confirm the UAT above? This pull has been sitting out here a while now

**Comment by andersdd** [comment]
@gpazo Confirmed user described scenario and working on a fix now.

**Comment by Panduk** [comment]
Performed UAT: Issue is fixed.

File: foundation-database/public/functions/releasenumber.sql [hunk 1]
```diff
@@ -28,16 +29,26 @@ BEGIN
...
+
+    EXECUTE _select INTO _test;
 
-    -- check if order seq has been incremented past the given order number
-    -- S/O code reads: IF ((_test - 1) <> pSoNumber) THEN
-    -- but the following /should/ address bug 4020 (can't reproduce it to test)
-    IF (FOUND AND ((_test - 1) > pnumber)) THEN
+    -- Check if order seq has been incremented past the given order number
+    -- Problem occurred with open but not saved orders (i.e. open by two or more people
+    -- at the same time) so need to check for that as well. Issues #4060 and #24717
```
Comment (gilmoskowitz) [inline foundation-database/public/functions/releasenumber.sql]: the comment should refer to `4020`, not `4060`

**Pull request merged**